File: android/src/main/java/com/reactnativecloudpayments/GooglePayModule.java
package com.reactnativecloudpayments;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.BaseActivityEventListener;
import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.module.annotations.ReactModule;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.wallet.IsReadyToPayRequest;
import com.google.android.gms.wallet.PaymentDataRequest;
import com.google.android.gms.wallet.PaymentData;
import com.google.android.gms.wallet.PaymentsClient;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.wallet.AutoResolveHelper;
import com.google.android.gms.wallet.WalletConstants;

import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONArray;

import android.app.Activity;
import android.content.Intent;
import androidx.annotation.NonNull;

import java.util.Optional;
import java.lang.String;

@ReactModule(name = GooglePayModule.NAME)
public class GooglePayModule extends ReactContextBaseJavaModule  {
  public static final String NAME = "GooglePay";
  private RequestGooglePay requestGooglePay;
  private int ENVIRONMENT_RUNNING;
  private int REQUEST_CODE = 13; // Arbitrarily-picked constant integer you define to track a request for payment data activity.

  private ActivityEventListener activityEventListener = new BaseActivityEventListener() {
    @Override
    public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
      if (requestCode == REQUEST_CODE) {
        if (resultCode == Activity.RESULT_OK) {
          PaymentData paymentData = PaymentData.getFromIntent(data);
          getSuccessPaymentData(paymentData);
        }
      }
    }
  };

  public GooglePayModule(ReactApplicationContext context) {
    super(context);
    this.requestGooglePay = new RequestGooglePay();
    context.addActivityEventListener(this.activityEventListener);
  }

  private void getSuccessPaymentData (PaymentData paymentData) {
    try {
      EventEmitter eventEmitter = new EventEmitter();
      String paymentInformation = paymentData.toJson();
      JSONObject paymentMethodData = new JSONObject(paymentInformation).getJSONObject("paymentMethodData");
      String token = paymentMethodData.getJSONObject("tokenizationData").getString("token");
      eventEmitter.sendEvent(getReactApplicationContext(), "listenerCryptogramCard", token);
    } catch(JSONException e) {
      e.printStackTrace();
    }
  }

  @ReactMethod
  public void setGatewayTokenSpecification(String gateway, String gatewayMerchantId){
    requestGooglePay.setGatewayTokenSpecification(gateway, gatewayMerchantId);
  }

  @ReactMethod
  public void setPaymentNetworks(ReadableArray listPaymentNetwork) {
    JSONArray paymentNetworks = new JSONArray();
    for (int i = 0; i < listPaymentNetwork.size(); i++) {
      paymentNetworks.put(listPaymentNetwork.getString(i));
    }
    requestGooglePay.setPaymentNetworks(paymentNetworks);
  }


  @ReactMethod
  public void canMakePayments(Promise promise) {
    Activity activity = getCurrentActivity();
    PaymentsClient paymentsClient = requestGooglePay.createPaymentsClient(activity, ENVIRONMENT_RUNNING);
    final Optional<JSONObject> isReadyToPayJson = requestGooglePay.getIsReadyToPayRequest();

    if (!isReadyToPayJson.isPresent()) {
      promise.resolve(false);
      return;
    }

    IsReadyToPayRequest request = IsReadyToPayRequest.fromJson(isReadyToPayJson.get().toString());
    Task<Boolean> task = paymentsClient.isReadyToPay(request);

    task.addOnCompleteListener(activity, new OnCompleteListener<Boolean>() {
      @Override
      public void onComplete(@NonNull Task<Boolean> task) {
        if (task.isSuccessful()) {
          promise.resolve(task.getResult());
        } else {
          promise.resolve(false);
        }
      }
    });
  }

  @ReactMethod
  public void setRequestPay(String countryCode, String currencyCode, String merchantName, String merchantId) {
    requestGooglePay.setRequestPay(countryCode, currencyCode, merchantName, merchantId);
  }

  @ReactMethod
  public void setProducts(ReadableMap product) {
    String price = product.getString("price");
    requestGooglePay.setProducts(price);
  }

  @ReactMethod
  public void setEnvironment(int environmentRunning) {
    this.ENVIRONMENT_RUNNING = environmentRunning;
  }

  @ReactMethod
  public void openGooglePay() {
    Activity activity = getCurrentActivity();
    PaymentsClient paymentsClient = requestGooglePay.createPaymentsClient(activity, ENVIRONMENT_RUNNING);

    String paymentDataRequest = requestGooglePay.getPaymentDataRequest().toString();
    PaymentDataRequest request = PaymentDataRequest.fromJson(paymentDataRequest);

    if (request != null) {
      AutoResolveHelper.resolveTask(paymentsClient.loadPaymentData(request), activity, REQUEST_CODE);
    }
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }
}

File: android/src/main/java/com/reactnativecloudpayments/GooglePay/GooglePayModule.kt
package com.reactnativecloudpayments

import android.app.Activity
import android.content.Intent
import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.wallet.*
import org.json.JSONObject


@ReactModule(name = GooglePayModule.MODULE_NAME)
class GooglePayModule(reactContext: ReactApplicationContext): ReactContextBaseJavaModule(reactContext), ActivityEventListener {
  companion object {
    const val MODULE_NAME: String = "GooglePayModule"
    private var REQUEST_CODE_PAYMENT = 13  // код выполнения оплаты для отслеживания выполнения операции
  }

  override fun getName() = GooglePayModule.MODULE_NAME;

  lateinit var googlePayRequest: GooglePayRequest;
  lateinit var paymentsClient: PaymentsClient;
  lateinit var countryCode: String;
  lateinit var currencyCode: String;

  override fun onActivityResult(activity: Activity?, requestCode: Int, resultCode: Int, data: Intent?) {
    if (requestCode == REQUEST_CODE_PAYMENT) {
      if (resultCode == Activity.RESULT_OK) {
        data?.let { intent ->
          val paymentData = PaymentData.getFromIntent(intent);

          val paymentMethodData: JSONObject = JSONObject(paymentData?.toJson() as String).getJSONObject("paymentMethodData")
          val tokenGP = paymentMethodData.getJSONObject("tokenizationData").getString("token") as String;

          EventEmitter.sendEvent(reactApplicationContext, "listenerCryptogramCard",  tokenGP)
        }
      }
    }
  }

  override fun onNewIntent(intent: Intent?) = Unit

  @ReactMethod
  fun initialization(initialData: ReadableMap, paymentNetworks: ReadableArray) {
    val initialDataParsed = GooglePayMethodData(initialData);
    val paymentNetworksList = paymentNetworks.toArrayList();

    countryCode = initialDataParsed.countryCode;
    currencyCode = initialDataParsed.currencyCode;

    googlePayRequest = GooglePayRequest(
      initialDataParsed.merchantName,
      initialDataParsed.googlePayMerchantId,
      initialDataParsed.gateway,
      paymentNetworksList
    );

    val activity = currentActivity as Activity;

    paymentsClient = googlePayRequest.createPaymentsClient(activity, initialDataParsed.environmentRunning);

    reactApplicationContext.addActivityEventListener(this)
  }

  @ReactMethod
  fun setProducts(totalPrice: String) {
    googlePayRequest.setTransactionInfo(countryCode, currencyCode, totalPrice);
  }

  @ReactMethod
  fun canMakePayments(promise: Promise) {
    val isReadyToPayJson = googlePayRequest.isReadyToPayRequest() ?: return
    val request = IsReadyToPayRequest.fromJson(isReadyToPayJson.toString());
    val task = paymentsClient.isReadyToPay(request)

    task.addOnCompleteListener { completedTask ->
      try {
        if(completedTask.getResult(ApiException::class.java)) {
          promise.resolve(true);
        } else {
          promise.resolve(false);
        }
      } catch (exception: ApiException) {
        exception.printStackTrace()
        promise.resolve(false);
      }
    }
  }

  @ReactMethod
  fun open() {
    val paymentDataRequest: String = googlePayRequest.getPaymentDataRequest().toString()
    val request = PaymentDataRequest.fromJson(paymentDataRequest) // конвертация из json в класс дата

    AutoResolveHelper.resolveTask(paymentsClient.loadPaymentData(request), currentActivity!!, REQUEST_CODE_PAYMENT)
  }
}
